sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/EncryptionCosmosAsyncDatabase.java

```diff
@@ -0,0 +1,392 @@
+     * <p>
+     * After subscription the operation will be performed. The {@link CosmosPagedFlux} will
+     * contain one or several feed response of the read cosmos client encryption keys. In case of
+     * failure the {@link CosmosPagedFlux} will error.
+     *
+     * @return a {@link CosmosPagedFlux} containing one or several feed response pages of the
+     * read cosmos client encryption keys or an error.
+     */
+    public CosmosPagedFlux<CosmosClientEncryptionKeyProperties> readAllClientEncryptionKeys() {
+        return this.cosmosAsyncDatabase.readAllClientEncryptionKeys();
+    }
+
+    /**
+     * Creates a client encryption key after subscription the operation will be performed. The
+     * {@link Mono} upon successful completion will contain a single resource
+     * response with the created client encryption key. In case of failure the {@link Mono} will
+     * error.
+     *
+     * @param clientEncryptionKeyId     Client Encryption Key id.
+     * @param encryptionAlgorithm       Encryption Algorithm.
+     * @param encryptionKeyWrapMetadata EncryptionKeyWrapMetadata.
+     * @return an {@link Mono} containing the single resource response with the
+     * created cosmos client encryption key or an error.
+     */
+    public Mono<CosmosClientEncryptionKeyResponse> createClientEncryptionKey(String clientEncryptionKeyId,
+                                                                             String encryptionAlgorithm,
+                                                                             EncryptionKeyWrapMetadata encryptionKeyWrapMetadata) {
+        if (StringUtils.isEmpty(clientEncryptionKeyId)) {
+            return Mono.error(new IllegalArgumentException("clientEncryptionKeyId is null or empty"));
+        }
+
+        if (StringUtils.isEmpty(encryptionAlgorithm)) {
+            return Mono.error(new IllegalArgumentException("encryptionAlgorithm is null or empty"));
+        }
+
+        EncryptionKeyStoreProvider encryptionKeyStoreProvider =
+            this.encryptionCosmosAsyncClient.getEncryptionKeyStoreProvider();
+        try {
+            KeyEncryptionKey keyEncryptionKey = KeyEncryptionKey.getOrCreate(encryptionKeyWrapMetadata.name,
+                encryptionKeyWrapMetadata.value, encryptionKeyStoreProvider, false);
+            ProtectedDataEncryptionKey protectedDataEncryptionKey =
+                new ProtectedDataEncryptionKey(clientEncryptionKeyId, keyEncryptionKey);
+            byte[] wrappedDataEncryptionKey = protectedDataEncryptionKey.getEncryptedValue();
+            CosmosClientEncryptionKeyProperties clientEncryptionKeyProperties =
+                new CosmosClientEncryptionKeyProperties(clientEncryptionKeyId, encryptionAlgorithm,
+                    wrappedDataEncryptionKey, encryptionKeyWrapMetadata);
+            return this.cosmosAsyncDatabase.createClientEncryptionKey(clientEncryptionKeyProperties);
+        } catch (Exception ex) {
```

Comment:
who can throw here? isn't this block all async?

Reply:
This is on  AzureKeyVaultKeyStoreProvider encryptionKeyStoreProvider = new AzureKeyVaultKeyStoreProvider(tokenCredentials), outside cosmos